File: App/src/Adaptive.ReactiveTrader.Client/ServiceClients/Execution/ExecutionServiceClient.cs
using System;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Transport;
using Adaptive.ReactiveTrader.Shared;
using Adaptive.ReactiveTrader.Shared.Execution;
using Adaptive.ReactiveTrader.Shared.Extensions;
using Microsoft.AspNet.SignalR.Client;

namespace Adaptive.ReactiveTrader.Client.ServiceClients.Execution
{
    class ExecutionServiceClient : IExecutionServiceClient
    {
        private readonly IConnectionProvider _connectionProvider;

        public ExecutionServiceClient(IConnectionProvider connectionProvider)
        {
            _connectionProvider = connectionProvider;
        }

        public IObservable<TradeDto> Execute(TradeRequestDto tradeRequest)
        {
            return (from connection in _connectionProvider.GetActiveConnection().Take(1) 
                from trade in ExecuteForConnection(connection.ExecutionHubProxy, tradeRequest)
                select trade)
                .CacheFirstResult();
        }

        private IObservable<TradeDto> ExecuteForConnection(IHubProxy executionHubProxy, TradeRequestDto tradeRequestDto)
        {
            return
                Observable.FromAsync(
                    () => executionHubProxy.Invoke<TradeDto>(ServiceConstants.Server.Execute, tradeRequestDto));
        } 
    }
}

File: src/Adaptive.ReactiveTrader.Client.JS/ServiceClients/ExecutionServiceClient.ts
class ExecutionServiceClient implements IExecutionServiceClient {
    private _connection: IConnection;

    constructor(connection: IConnection) {
        this._connection = connection;
    }

    public execute(tradeRequest: TradeRequestDto): Rx.Observable<TradeDto> {
        return this.executeForConnection(tradeRequest, this._connection.executionHubProxy);
    }

    public executeForConnection(tradeRequest: TradeRequestDto, executionHub: HubProxy): Rx.Observable<TradeDto> {
        return Rx.Observable.create<TradeDto>(observer=> {
            executionHub.invoke("Execute", tradeRequest)
                .done(trade=> observer.onNext(trade))
                .fail(error=> observer.onError(error));

            return Rx.Disposable.empty;
        })
        .take(1)
        .timeout(2000);
        // TODO cachefirstresult
    }
}